src/plugins/index_pattern_management/public/components/create_index_pattern_wizard/components/step_index_pattern/step_index_pattern.tsx

```diff
@@ -40,29 +39,56 @@ import { IndicesList } from './components/indices_list';
 import { Header } from './components/header';
 import { context as contextType } from '../../../../../../kibana_react/public';
 import { IndexPatternCreationConfig } from '../../../../../../../plugins/index_pattern_management/public';
-import { MatchedIndex } from '../../types';
+import { MatchedItem, ResolveIndexResponseItemDataStream } from '../../types';
 import { IndexPatternManagmentContextValue } from '../../../../types';
 
 interface StepIndexPatternProps {
-  allIndices: MatchedIndex[];
-  isIncludingSystemIndices: boolean;
+  allIndices: MatchedItem[];
   indexPatternCreationType: IndexPatternCreationConfig;
-  goToNextStep: (query: string) => void;
+  goToNextStep: (query: string, timestampField?: string) => void;
   initialQuery?: string;
+  showSystemIndices: boolean;
 }
 
 interface StepIndexPatternState {
-  partialMatchedIndices: MatchedIndex[];
-  exactMatchedIndices: MatchedIndex[];
+  partialMatchedIndices: MatchedItem[];
+  exactMatchedIndices: MatchedItem[];
   isLoadingIndices: boolean;
   existingIndexPatterns: string[];
   indexPatternExists: boolean;
   query: string;
   appendedWildcard: boolean;
   showingIndexPatternQueryErrors: boolean;
   indexPatternName: string;
+  isIncludingSystemIndices: boolean;
 }
 
+export const canPreselectTimeField = (indices: MatchedItem[]) => {
+  const preselectStatus = indices.reduce(
+    (
+      { canPreselect, timeFieldName }: { canPreselect: boolean; timeFieldName?: string },
+      matchedItem
+    ) => {
+      const dataStreamItem = matchedItem.item as ResolveIndexResponseItemDataStream;
```

Comment:
nit: Would be nice to use typeguards to differentiate between the different possible `matchedItem.item` types so that you don't need casting here.

Or perhaps make `MatchedItem` generic so that you can optionally specify an item type if it is known at compile time:
```ts
type MatchedItemTypes =
    | ResolveIndexResponseItemIndex
    | ResolveIndexResponseItemAlias
    | ResolveIndexResponseItemDataStream;

export interface MatchedItem<I extends MatchedItemTypes = MatchedItemTypes> {
  name: string;
  tags: Tag[];
  item: I;
}

// usage:
type MatchedDataStreamItem = MatchedItem<ResolveIndexResponseItemDataStream>;
export const canPreselectTimeField = (indices: MatchedDataStreamItem[]) => {}
```

Reply:
ah, thats much better! That code was annoying me.

Reply:
I just made the MatchedItem['item'] definition more flexible. I think the larger problem is that I'm trying to determine a specific type from a general type. It seems that typescript will allow me to write a function to do it but that seems like overkill just to check a single field. The best fix would be to track Data source items in their own array but I'd rather not make that change at this stage.